keps/sig-auth/3325-self-user-attributes-review-api/README.md

```diff
@@ -0,0 +1,357 @@
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Status is filled in by the server with the user attributes.
+	Status SelfUserAttributesReviewStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+```
+```go
+type SelfUserAttributesReviewStatus struct {
+	// User attributes of the current user.
+	// +optional
+	UserInfo UserInfo `json:"userInfo,omitempty" protobuf:"bytes,1,opt,name=userInfo"`
+}
+
+type UserInfo struct {
+	Name   string              `json:"name" protobuf:"bytes,1,opt,name=name"`
+	UID    string              `json:"uid" protobuf:"bytes,2,opt,name=uid"`
+	Groups []string            `json:"groups" protobuf:"bytes,1,opt,name=groups"`
+	Extra  map[string][]string `json:"extra" protobuf:"bytes,1,opt,name=extra"`
+}
+```
+
+On receiving a request, the Kubernetes API server fills the status with the user attributes and returns it to the user.
+
+Request URL:
+```
+GET /apis/authentication.k8s.io/v1alpha1/selfuserattributesreview
+```
+Response example:
+
+```json
+{
+  "apiVersion": "authentication.k8s.io/v1alpha1",
+  "kind": "SelfUserAttributesReview",
+  "status": {
+    "name": "jane.doe",
+    "uid": "b6c7cfd4-f166-11ec-8ea0-0242ac120002",
+    "groups": ["viewers", "editors"],
+    "extra": {
+      "provider_id": "token.company.dev"
+    }
+  }
+}
+```
+
+User attributes are known at the moment of accessing the rest API endpoint and can be extracted from the request context.
+
+### RBAC
+
+RBAC rules to grant access to this API should be present in the cluster by default.
+It is implied that the `system:basic-user` cluster role will be extended to the following:
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  annotations:
+    rbac.authorization.kubernetes.io/autoupdate: "true"
+  creationTimestamp: null
+  labels:
+    kubernetes.io/bootstrapping: rbac-defaults
+  name: system:basic-user
+rules:
+- apiGroups:
+  - authorization.k8s.io
+  resources:
+  - selfsubjectaccessreviews
+  - selfsubjectrulesreviews
+  verbs:
+  - create
+- apiGroups:
+  - authentication.k8s.io
+  resources:
+  - selfuserattributesreview
+  verbs:
+  - create
+```
+
+This API is enabled by default and can be disabled by the following kube-apiserver flag (along with the TokenReview API).
+```
+--runtime-config=authentication.k8s.io/v1=false
+```
+
+### Test Plan
+
+Unit tests covering:
+
+1. Request returns all user attributes
+2. Request returns some user attributes
+3. Request with a status returns overridden fields
+
+Integration test covering:
+
+1. Successful authentication through a simple authenticator, e.g., token or certificate authenticator
+2. Successful authentication through a complicated authenticator, e.g., webhook or authentication proxy authenticator
+3. Failed authentication
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit and integration tests completed and enabled
```

Comment:
Is the plan to introduce this API in v1alpha1 first?

Reply:
Yes, that is the plan. I left a note about it. Mo [agreed](https://kubernetes.slack.com/archives/C0EN96KUY/p1655868809436129?thread_ts=1655497138.347379&cid=C0EN96KUY).